internal/pkg/fleetapi/action.go

```diff
@@ -538,99 +540,54 @@ type Actions []Action
 
 // UnmarshalJSON takes every raw representation of an action and try to decode them.
 func (a *Actions) UnmarshalJSON(data []byte) error {
-	var responses []FleetAction
-	if err := json.Unmarshal(data, &responses); err != nil {
+	var typeUnmarshaler []struct {
```

Comment:
Nit: couldn't we define the `typeUnmarshaler` as a single named struct containing only `ActionType` ?
We can then unmarshal a slice of them from the `json.Unmarshal()` below...

(Disclaimer: the part below only works for `json` package so if we need also yaml unmarshaling it's not applicable)
Another way could be to have a delayed parsing of `Data` using [`json.RawMessage`](https://pkg.go.dev/encoding/json#RawMessage) by defining the struct as (forgive my formatting)
```golang
struct {
		ActionType string `json:"type,omitempty" yaml:"type,omitempty"`
		...
		Data json.RawMessage `json:"data,omitempty"`
	}
```
and then unmarshaling to the correct type once type is known